File: Merge_sort.py
def merge(l,mid,h):
	n1 = mid-l+1
	n2 = h - mid
	p=[0]*(n1)
	q=[0]*(n2)
	for i in range(n1):
		p[i] = L[l+i]
	for i in range(n2):
		q[i] = L[mid+1+i]
	i=0
	j=0
	k = l
	while i<n1 and j<n2:
		if p[i]<=q[j]:
			L[k] = p[i]
			i += 1
		else:
			L[k] = q[j]
			j += 1
		k += 1	
	while i<n1:
		L[k] = p[i]
		i += 1
		k += 1
	while j<n2:
		L[k] = q[j]
		j += 1
		k += 1

def merge_sort(l,h):
	if l<h:
		mid = int((l+(h-1))/2)
		merge_sort(l,mid)
		merge_sort(mid+1,h)
		merge(l,mid,h)


L = [50,40,30,20,10]
n = len(L)
merge_sort(0,n-1)
print(L)

File: java/merge_sort.java
import java.util.Scanner;

class merge_sort
{
    void merge(int[] A,int m,int l,int r)
    {
        int p = m-l+1;
        int q = r-m;
        int[] B = new int[p];
        int[] C = new int[q];

        for(int i=0;i<p;i++)
            B[i] = A[l+i];
        for(int j=0;j<q;j++)
            C[j] = A[m+j+1];

        int i=0,j=0,k=l;
        while(i<p && j<q)
        {
            if(B[i] <= C[j])
                A[k++] = B[i++];
            else
                A[k++] = C[j++];
        }
        while(i<p)
            A[k++] = B[i++];
        while(j<q)
            A[k++] = C[j++];
    }

    void Sort(int[] A,int l,int r)
    {
        if(l < r)
        {
            int m = (l+r)/2;    //Brackets compulsory...
            Sort(A,l,m);
            Sort(A,m+1,r);
            merge(A,m,l,r);
        }
    }

    public static void main(String[] args)
    {
        merge_sort ob = new merge_sort();
        Scanner jin = new Scanner(System.in);
        System.out.print("Enter n value : ");
        int n = jin.nextInt();
        System.out.print("Enter elements : ");
        int[] A = new int[n];
        for(int i=0;i<n;i++)
            A[i] = jin.nextInt();
        ob.Sort(A,0,n-1);

        System.out.print("Sorted array : ");
        for(int i=0;i<n;i++)
            System.out.print(A[i]+" ");
        System.out.println();
    }
};
